src/eve/pattern_matching.py

```diff
@@ -0,0 +1,139 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import inspect
+from dataclasses import dataclass
+from types import ModuleType
+from typing import Any, Iterator
+
+
+@dataclass(frozen=True)
+class ObjectPattern:
+    cls: type
+    attrs: dict[str, Any]
```

Comment:
I'm not too much into using `fields` as what is stored here applies to general objects for which the common name is attribute (correct me if I'm wrong here). How about `attributes`?

Reply:
I proposed `fields` because it is now a widespread way to call plain data attributes of value classes (both  `dataclasses` and `pydantic` models use this convention), but `attributes` is also fine for me. The main point is to not to use `attrs`